### ISSUE TITLE
`connected_to` isn't using specified database on first request

### ISSUE THREAD

**Issue by colonelschnapps** [rails/rails#38548]
TLDR: How should `connects_to` and `connected_to` be used together? I haven't been able to find many examples of this out there, and believe that either I'm not using these as intended, or different use cases of these things in different combinations haven't been fully fleshed out yet.

Given a simple test project with `primary` and `secondary` keys in my db config, and:

```
class ApplicationRecord < ActiveRecord::Base
  self.abstract_class = true
  connects_to database: { writing: :primary, reading: :primary }
end
```
```
class EmployeesController < ApplicationController
  def index
    ActiveRecord::Base.connected_to(database: :secondary) do
      @employees = Employee.all
    end
    render json: @employees
  end
end
```

The db specified under `secondary` is a copy of the one under `primary`, except that the primary db has no employee data, and the secondary db has one test record.

Once the server is completely finished starting (so as to eliminate any intermediate initialization issues), I make a curl request to the above endpoint, and no data is returned.

I check my Postgres log, and can see that the `SELECT` query generated above is hitting the *primary* database (because I'm including the db name in my pg log output).

I then make a *second* request, and that and subsequent requests will return the single test record.

So it's the initial request that is hitting `primary`, while subsequent ones use `secondary` as specified in the block.

**If I comment out `connects_to()` in the `ApplicationRecord` class, it works as expected, returning the record on the first try.**

Am I using these two things together incorrectly? From the doc, I assumed that `connected_to()` would supersede everything else:

> For cases where you may want to connect to a database outside of the model, you can use connected_to with a database argument.

But it seems to be taking the `connects_to` configuration into account, at least for the first request (only).

I then tried creating dedicated writing and reading roles and specifying those since the above roles didn't actually exist, and modifying the block call to `ActiveRecord::Base.connected_to(database: { emp_reader: :secondary })`, but at that point started receiving empty results every time, because it was only accessing the `primary` database (as confirmed in my logs).

I've spent hours debugging this and trying different combinations of things, and at times am seeing unexpected and possibly unanticipated behavior.

I also tried the 6.1.0 alpha version based on [this](https://github.com/rails/rails/issues/38332), but had the same results.

Is there an example that shows how one would use a primary database with a replica, and a secondary or others with different databases, that may be called manually as in the block form above?

**Comment by eileencodes** [comment]
`connects_to` is basically establish connection, but you can establish more. `connected_to` is, as the docs say, for switching connections. The `database` key on `connected_to` is deprecated because it's not supposed to be used during requests and that's what everyone is using it for - it will always establish a new connection for every call. You should use it like this:

```
class ApplicationRecord < ActiveRecord::Base
  self.abstract_class = true
  connects_to database: { writing: :primary, reading: :primary }
end
```

```
class EmployeesController < ApplicationController
  def index
    ActiveRecord::Base.connected_to(role: :reading) do
      @employees = Employee.all
    end
    render json: @employees
  end
end
```

If you're looking for sharding Rails 6.0 doesn't support that. I opened a PR to support it which I will hopefully merge next week.

**Comment by colonelschnapps** [comment]
@eileencodes Thanks for the quick response.

In the example above, which database is `connected_to` connected to? 😄 

I'm trying to use two different databases here, and it appears this is only connecting to one? (`reading` is associated with `primary`, not `secondary`, no?)

**Comment by eileencodes** [comment]
Pretend the `database` key on `connected_to` doesn't exist. It was a mistake for me to implement it and I have serious regrets about doing that. It was supposed to be for one-off connections for scripts that use a different type of connection, like a replica that allows slow connections. But everyone is using it in requests and it should absolutely not be used in requests because it will establish a new connection every time it's called. 😅 

> In the example above, which database is connected_to connected to? 

`connected_to` switches _roles_ not databases. Once you switch `roles` Rails looks up the connection specification name based on the record you try to call. So when you call Employee it will use the ActiveRecord::Base connection identified by the reading role (role is a ConnectionHandler object). 

> What if there were only one role, and two databases

I _think_ you're talking about horizontal sharding here, and like I said Rails 6.0 doesn't support sharding. When I merge #38531 6.1/master will support it, but 6.0 never will. In 6.0 there's no way for 1 model to connect to 2 databases on the same role. 6.0 is specifically designed to be 1 db per role per parent model.

Role (writing) -> ApplicationRecord -> :primary
Role (reading) -> ApplicationRecord -> :primary_replica

So in your case for you to use secondary on reading it needs to be a replica and the database names need to be the same. Here is what your app would look like

```
# config/database.yml
development:
  primary:
    database: my_db
  secondary:
    database: my_db # must match the same db in primary, can have different users though
    replica: true
```

```
# app/models/application_record.rb
class ApplicationRecord < ActiveRecord::Base
  self.abstract_class = true

  connects_to database: { writing: :primary, reading: :secondary }
end
```

```
# app/controllers/employees_controller.rb
class EmployeesController < ApplicationController
  def index
    ActiveRecord::Base.connected_to(role: :reading) do
      @employees = Employee.all # reads from secondary
    end
    render json: @employees
  end
end
```

**Comment by colonelschnapps** [comment]
Or maybe I should phrase more generally. If I have two databases, one of which remains primary and is used throughout the app, and a second that I want to use in only one call or endpoint, for example, what's the best way to handle that?

Should I look to the older method of using `establish_connection()`? (But this *is* the 'multiple databases' feature...)

**Comment by colonelschnapps** [comment]
(crossed responses) The above is really what I'm after. If I want to connect to a second database in a limited context, can I make this work?

I did read the guide and that was the sticking point for me--that the database is the same in both configurations/keys in the example above above.

**Comment by eileencodes** [comment]
> If I have two databases, one of which remains primary and is used throughout the app, and a second that I want to use in only one call or endpoint, for example, what's the best way to handle that?

> If I want to connect to a second database in a limited context, can I make this work?

To be honest I have a hard time imagining what is a use case for this in a request. In the context of a request you don't want to be just establishing new connections on the fly. The way connections work in Rails, this would cause the existing connection to get overwritten if it has the same parent model. So this is why I ended up deprecating the database key on `connected_to` - the behavior during a request is kind of dangerous. Even with sharding the intent is to establish a connection on application boot not on the fly.

`establish_connection` will work in that it will establish a new connection but you'd be overwriting the existing one because the connection owner class, ActiveRecord::Base is the same. It has the same problem as `connected_to` with the `database` kwarg.

**Comment by eileencodes** [comment]
I don't have time to try this out right now but this _might_ work?

```
# config/database.yml
development:
  primary:
    database: my_db
  secondary:
    database: my_secondary_db
```

```
# app/models/application_record.rb
class ApplicationRecord < ActiveRecord::Base
  self.abstract_class = true

  connects_to database: { writing: :primary, some_other_name_that_is_not_reading: :secondary }
end
```

```
# app/controllers/employees_controller.rb
class EmployeesController < ApplicationController
  def index
    ActiveRecord::Base.connected_to(role: :some_other_name_that_is_not_reading) do
      @employees = Employee.all # reads from secondary
    end
    render json: @employees
  end
end
```

**Comment by colonelschnapps** [comment]
Makes sense, there are connection and pooling considerations, so I don't want to mess with that. 🏊 

I figured going back to the more general use case would help avoid some of the implementation-specific details: "Say I have two databases..."

I'm going to experiment some more, but you also gave me another example above, so I will try and report back. ✏️

**Comment by colonelschnapps** [comment]
It doesn't seem to be working--no record result, pg logs shows primary db called--and I'm trying to figure out why now. (This was one of the different setups I've tried along the way before posting.)

It appears to be returning the correct `ActiveRecord::ConnectionAdapters::ConnectionHandler`, the one associated with the second role (which actually is a read-only role, permitted only on the secondary db).

It's also returning a `ConnectionPool` whose `ConnectionSpecification` contains the correct config: `{:adapter=>"postgresql", :encoding=>"unicode", :host=>"localhost", :pool=>5, :database=>"emp_replica", :user=>"emp_reader"}`.

One thing I notice is that the `ConnectionSpecification`'s `name` is 'primary,' not 'secondary.' So its config is correct, but not its name. I'm trying to figure out why now...

**Comment by eileencodes** [comment]
> It doesn't seem to be working--no record result, pg logs shows primary db called--and I'm trying to figure out why now. (This was one of the different setups I've tried along the way before posting.)

So you have 2 dbs and expect the same data. _Something_ has to copy the data from one to the other, Rails doesn't do that for you.

> One thing I notice is that the ConnectionSpecification's name is 'primary,' not 'secondary.' So its config is correct, but not its name. I'm trying to figure out why now...

ConnectionSpecification will never be secondary, it points to the class name, not the spec name on the database config. It's [primary in 6.0](https://github.com/rails/rails/blob/6-0-stable/activerecord/lib/active_record/connection_handling.rb#L209-L215) and will be [ActiveRecord::Base on 6.1](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/connection_handling.rb#L180-L185) because it never made sense to me to give AR Base a different name.

**Comment by colonelschnapps** [comment]
Nah, it's all described above (large block of text, I know)--this is all simplified for the test, and constructed manually.

`primary` has an empty `employees` table, while `secondary` has one row, so that I can differentiate between dbs in the response. (I could create a different row(s) too, but this works.)

**Comment by colonelschnapps** [comment]
So `curl localhost:3000/employees` returns '[]', and my (pre-emptied) db log shows:

`2020-02-22 09:23:38 PST emp_development LOG:  duration: 0.006 ms  execute <unnamed>: SELECT "employees".* FROM "employees"`

But if I comment out the `connects_to` in `ApplicationRecord`, and restore the (bad) version of `connected_to`, as in: `ActiveRecord::Base.connected_to(database: :secondary) do`, it returns the one result from the `secondary` (non-empty) db:

`[{"id":2,"first_name":"George","last_name":"Washington","created_at":"2020-02-21T21:13:58.206Z","updated_at":"2020-02-21T21:13:58.206Z"}]`

**Comment by eileencodes** [comment]
I'm having a really hard time following this and I can't spend anymore time on this today, sorry. If you can write a demo app that would help and I can try to look on Monday.

**Comment by courtsimas** [comment]
I just wanted to speak into this and say this was immensely helpful. I was trying to move some data over in a rake task from an old table to a new table (from one db to another) - named the same  (same model/table), and I couldn't get the switch to happen. I was using the database key, and that wasn't working. 

Once I thought how to approach this without using that, using roles only, I made a third role for the table that was moving over, and that role connected to the other db. 

That made this super easy. I'm sure this is old news now but to have this documented somewhere would be really helpful...

**Comment by eileencodes** [comment]
> I'm sure this is old news now but to have this documented somewhere would be really helpful...

I just realized the `database` kwarg is still in the guides so I'll remove it. The database kwarg for `connected_to` is already deprecated in 6.1 and will be removed in 6.2 without replacement.

**Comment by courtsimas** [comment]
@eileencodes thank you... Also great work on this amazing rails feature!
